### PR TITLE
Added fix so that staging data still gets cleaned up even if collectOutput fails

### PR DESCRIPTION
**Pull request by sahilTakiar**
The problem is that line 367 of AbstractJobLauncher.addWorkUnit() adds each WorkUnit from Source.getWorkUnits() to the jobState.taskStates Map. So the Map contains all the pre-runtime values of the WorkUnits. However, on line 308 of Fork.buildWriter(), the value of ConfigurationKeys.WRITER_FILE_PATH is set if it is not currently present in the WorkUnit. Fork.buildWriter() is only called during the runtime of each WorkUnit. So the value of ConfigurationKeys.WRITER_FILE_PATH never makes it into the JobState object. So, the JobLauncherUtils.cleanStagingData() method won’t do anything, as the WorkUnit it is looking at doesn’t have a value for  ConfigurationKeys.WRITER_FILE_PATH.

This pull request will ensure this doesn't happen again. If the WRITER_FILE_PATH does not exist in the config during cleanup, the code uses its default value to cleanup the data.

### CODE REVIEW

File: gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java [hunk 1]
```diff
@@ -71,27 +71,30 @@ public static String newMultiTaskId(String jobId, int sequence) {
...
-      String writerFilePath = taskState
+      String writerFilePath = state
           .getProp(ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_FILE_PATH, branches, i));
+
+      // If the WRITER_FILE_PATH has not been set, use the default value
       if (Strings.isNullOrEmpty(writerFilePath)) {
-        // The job may be cancelled before the task starts, so this may not be set.
-        continue;
+        writerFilePath =
+            state.getExtract().getOutputFilePath() + Path.SEPARATOR
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java]: We need a helper class to construct the staging directory path which should be used by this class and the Fork operator .

File: gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java [hunk 2]
```diff
@@ -71,27 +71,30 @@ public static String newMultiTaskId(String jobId, int sequence) {
...
 
       String stagingDirKey =
           ForkOperatorUtils.getPropertyNameForBranch(ConfigurationKeys.WRITER_STAGING_DIR, branches, i);
-      if (taskState.contains(stagingDirKey)) {
-        Path stagingPath = new Path(taskState.getProp(stagingDirKey), writerFilePath);
+      if (state.contains(stagingDirKey)) {
+        Path stagingPath = new Path(state.getProp(stagingDirKey), writerFilePath);
         if (fs.exists(stagingPath)) {
           logger.info("Cleaning up staging directory " + stagingPath.toUri().getPath());
           fs.delete(stagingPath, true);
```
Comment (chavdar) [inline gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java]: Is it possible to delete the entire extract staging directory? There are a lot of HDFS operations here and anyone can throw an exception which will leave the staging directory in an inconsistent state. Ideally, that directory is removed in the beginning of the job and the job should fail if the clean up of that directory fails.

File: gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java (hunk 1 above)
Comment (sahilTakiar) [inline gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java]: Added, its called `WriterUtils`.

File: gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java (hunk 2 above)
Comment (sahilTakiar) [inline gobblin-utility/src/main/java/gobblin/util/JobLauncherUtils.java]: I added code in `MRJobLauncher` to cleanup the staging and output directories of each `WorkUnit`.

**Comment by sahilTakiar** [comment]
New version to address comments. Also, made some changes to cleanup how writer output paths and directories are constructed. This pull request is what I have so far, but the tests don't all pass yet, so I'm working on that now.

Also, need to add more javadocs and tests.

**Comment by sahilTakiar** [comment]
I tried to rebase this pull request, but it didn't work, will make a new one.

**Pull request closed without merging**